Fix caching when HTMLDocument::countModules is used

Fedik opened the pull request:
Pull Request for Issue #35180 and #34568 .

### Summary of Changes
Fix modules caching when `$this->countModules('position', true)` is used in a template.


### Testing Instructions
Please follow #35180


### Actual result BEFORE applying this Pull Request
Some modules have broken style


### Expected result AFTER applying this Pull Request
All works 


### Documentation Changes Required
none

reference #19416

Fedik committed:
Fix caching when HTMLDocument::countModules is used

PhilETaylor commented on libraries/src/Helper/ModuleHelper.php:
```diff
@@ -165,6 +165,8 @@ public static function renderModule($module, $attribs = array())
 		// Render the module content
 		static::renderRawModule($module, $params, $attribs);
 
+		// Check for a "raw" style: render module content only.
+		// This style only for internal use, and should not be used within template markup.
```
CS needs /* */ for multi line comments

PhilETaylor commented on libraries/src/Document/Renderer/Html/ModuleRenderer.php:
```diff
@@ -84,14 +84,28 @@ public function render($module, $attribs = array(), $content = null)
+
+			$result = ModuleHelper::moduleCache($module, $params, $cacheparams);
+
+			// For a raw style set the content to the module, so it will not be rendered again next time, see ModuleHelper::renderRawModule
+			// It need to be done here because the cache controller does not keep reference to the module object
```
CS needs /* */ for multi line comments

PhilETaylor commented on libraries/src/Document/HtmlDocument.php:
```diff
@@ -695,6 +695,8 @@ public function countModules(string $positionName, bool $withContentOnly = false
 		{
 			if (empty($module->contentRendered))
 			{
+				// Use style "raw", that will render only the content of the module, and precache it, see ModuleHelper::renderRawModule
+				// This style only for internal use, and should not be used within template markup.
```
CS needs /* */ for multi line comments

PhilETaylor commented:
Testing instructions - MAKE SURE YOU CLEAR YOUR CACHE after applying this PR if you already have a warm PR. I didn't and I still saw a broken chrome until I did clear the cache manually, after that it was fine.

PhilETaylor commented:
I have tested this item :white_check_mark: successfully on a73cfb2809debc511ced8d0b0fe6945a159fa9ad<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/35225">issues.joomla.org/tracker/joomla-cms/35225</a>.</sub>

HLeithner committed:
Simplify module counting

HLeithner committed:
cs

HLeithner committed:
Rename parameter

Fedik commented on libraries/src/Helper/ModuleHelper.php:
```diff
@@ -218,7 +213,8 @@ public static function renderModule($module, $attribs = array())
 
-		if ($module === null || !isset($module->content))
+		// Don't apply chromes if we only need the content
+		if ($module === null || !isset($module->content) || !empty($attribs['contentOnly']))
```
@HLeithner I think this will make a bug, the `onRenderModule` will be called twice,
first time when countModule, and second time when render chrome style

now I remember why I made early return, originally

brianteeman replied:
and wasnt that the purpose of the original PR to prevent it happening twice?

HLeithner replied:
Would this be a problem?
I see it as a bigger problem if you don't call it for counting.

think of a plugin that remove the content if the content is '< div >< /div >'.

or a plugin that adds content when a content for a module is empty.

in both cases countModules would be wrong.

on the other side calling the trigger twice of the same module shouldn't be a problem because I can have the same module on multiple positions and the module and plugin would be called on every position if I'm not wrong.

HLeithner replied:
reverted

Fedik commented on libraries/src/Document/HtmlDocument.php:
```diff
@@ -695,7 +695,9 @@ public function countModules(string $positionName, bool $withContentOnly = false
-				$renderer->render($module, array('style' => 'raw'));
+				// Preserve content if loaded from cache
+				$module->content = $renderer->render($module, ['contentOnly' => true]);
+				$module->contentRendered = true;
```
hmhm, this flag already set by `renderRawModule`, 
it only may need to set when cache enabled, after ModuleHelper::moduleCache

HLeithner committed:
Removed local caching for module counting

Local caching for module count's could lead to diffrent results when count is be called before and after rendering a module position.

Fedik commented on libraries/src/Document/HtmlDocument.php:
```diff
@@ -693,10 +693,7 @@ public function countModules(string $positionName, bool $withContentOnly = false
-			{
-				$renderer->render($module, array('style' => 'raw'));
-			}
+			$renderer->render($module, ['contentOnly' => true]);
```
I still would do:
```
if (empty($module->contentRendered))
{
	$renderer->render($module, ['contentOnly' => true]);
}
```
To avoid extra call of renderer

HLeithner replied:
reverted

Fedik commented on libraries/src/Document/Renderer/Html/ModuleRenderer.php:
```diff
@@ -90,8 +90,12 @@ public function render($module, $attribs = array(), $content = null)
 
-			return ModuleHelper::moduleCache($module, $params, $cacheparams);
+			// It need to be done here because the cache controller does not keep reference to the module object
+			$module->content = ModuleHelper::moduleCache($module, $params, $cacheparams);
```
and set `$module->contentRendered = true` here, in the same reason, becasue cach controller does not keep reference

HLeithner replied:
done

HLeithner committed:
Disable trigger onRenderModule again for content only rendering

Fedik commented:
Ready for testing again.

@PhilETaylor please re test :)

maikol-ortigueira commented:
I have tested this item :white_check_mark: successfully on 6fad521d6a0b4c1d2bf9a83af60ca02bd94c77f0<br /><br />I have tested the patch and it works perfectly. I was indeed getting the issue in both Firefox 91.x and Chrome 92.0.4515.131.

I did the test on a local installation with the following configuration:

Server: Devilbox
Joomla version: 4.0.1-dev
PHP version: 7.4.13
Web server: Apache/2.4.5
Database version: Maria DB 10.3.27
Browser: Firefox 91.0 and Google Chrome 92.0.4515.131

Thank you very much.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/35225">issues.joomla.org/tracker/joomla-cms/35225</a>.</sub>

richard67 commented:
@PhilETaylor Could you test this one again? Thanks in advance.

richard67 commented:
@Fedik I think this fixes also issue #34568 , right?

PhilETaylor commented:
I have tested this item :white_check_mark: successfully on 6fad521d6a0b4c1d2bf9a83af60ca02bd94c77f0<br /><br />Seems to work. <hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/35225">issues.joomla.org/tracker/joomla-cms/35225</a>.</sub>

richard67 commented:
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/35225">issues.joomla.org/tracker/joomla-cms/35225</a>.</sub>

Fedik commented:
hm, yes, looks like

wilsonge commented on libraries/src/Helper/ModuleHelper.php:
```diff
@@ -165,7 +165,8 @@ public static function renderModule($module, $attribs = array())
 
-		if (!empty($attribs['style']) && $attribs['style'] === 'raw')
+		// Return early if only the content is required
+		if (!empty($attribs['contentOnly']))
```
But doesn't this mean we no longer call plugin events if contentOnly is set to true? I don't have a great feel of the code here - but something smells wrong.

HLeithner replied:
That's right but the fear for calling the event twice seems higher...

Fedik replied:
@wilsonge  "contentOnly" supposed to render only module content.
And our Module events called when rendering of module chrome, 
this way it works in old joomla and this true for j4, so here is no b/c.

And as @HLeithner already wrote, if we do `onModuleRender` before this check, then we will have double/tripple event trigger when someone use `HTMLDocuemnt::countModules()`

If need we can introduce another event that triggered from `ModuleHelper::renderRawModule()`, kind of `onAfterModuleContentRender`, but this for future. Not very good to add a new feature in 4.0.1 ;)
